Allow admin users to turn on automatic renewals if the subscription has a saved payment method

james-allan opened the pull request:
Fixes https://github.com/woocommerce/woocommerce-subscriptions/issues/4094
Fixes https://github.com/Automattic/woocommerce-subscriptions-core/issues/519

## Description

There's a feature in Woo Subscriptions that enables customers to switch between automatic and manual renewals. This is known as the auto renewal toggle.

| Auto-renewals off | Auto-renewals on |
|--------|--------|
| <img width="841" alt="Screenshot 2023-10-31 at 12 40 39 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/2b5c37f1-f026-4705-b861-caa6921bde09"> | <img width="830" alt="Screenshot 2023-10-31 at 12 41 08 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/6051b2ca-e0aa-4b30-b768-b277f9775364"> | 

When a customer turns off automatic renewals in this way, admin users are unable to switch it back to automatic renewals. This is because this toggle sets the `_requires_manual_renewal` flag which there's is no UI element for. 

This PR fixes that by adding a new checkbox on the edit subscription screen to allow them to turn automatic renewals back on.

https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/c4e2cea8-8d5b-4f27-9a9f-c570b964a9c7

<p align="center"><sup>The new "Auto renew" checkbox</sup></p>

## How to test this PR

1. Make sure you have Woo Subscriptions installed. 
2. Install this version of subscriptions-core (reminder: you can clone this repo and treat it like a plugin). 
3. Go to **WooCommerce → Settings → Subscriptions** 
4. Make sure the **Auto Renewal Toggle** setting is on.
5. Purchase a subscription using WooPayments, Stripe or any automatic payment method. 
6. Go to the My Account View Subscription page.
7. Click the Auto renewal toggle.
8. As admin go to **WooCommerce → Subscriptions** and edit the subscription.
   1. On `trunk` you'll notice that there's nothing you can do to turn automatic payments back on. Editing the billing details, doesn't enable you to set the auto renewal flag. 
   2. On this branch, you'll notice two things. 
       1. There are Subscription notes now added when the customer toggles auto renewal.
       2. If you edit the billing details, you will see a checkbox enabling you to turn on automatic renewals. (see video example above). 

<p align="center">
<img width="270" alt="Screenshot 2023-10-31 at 1 20 14 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/c336ceaa-ce94-4c7b-92cd-935781910c50">
</br>
<sup>Subscription notes</sup>
</p>

## Product impact 
<!-- What products will this PR ship in? -->

- [x] Added changelog entry (or does not apply)
- [ ] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [ ] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref
- [ ] <!-- 🚨 Deprecations 🚨 --> Added deprecated functions, hooks or classes to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0)

james-allan committed:
Add a subscription note when the customer turns off and on automatic renewals via their My Account page

james-allan committed:
Update capitalisation and use of colon to be consistant with other element lables

james-allan committed:
Add a checkbox enabling admin users to turn on automatic renewals

james-allan committed:
Add changelog for the new checkbox to re enable automatic renewals

james-allan committed:
Add inline comment

james-allan commented on includes/admin/meta-boxes/class-wcs-meta-box-subscription-data.php:
```diff
@@ -185,7 +185,7 @@ public static function output( $subscription ) {
 						}
 
-						echo '<p' . ( ( '' != $subscription->get_payment_method() ) ? ' class="' . esc_attr( $subscription->get_payment_method() ) . '"' : '' ) . '><strong>' . esc_html__( 'Payment Method', 'woocommerce-subscriptions' ) . ':</strong>' . wp_kses_post( nl2br( $subscription->get_payment_method_to_display() ) ); // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
+						echo '<p' . ( ( '' != $subscription->get_payment_method() ) ? ' class="' . esc_attr( $subscription->get_payment_method() ) . '"' : '' ) . '><strong>' . esc_html__( 'Payment method', 'woocommerce-subscriptions' ) . ':</strong>' . wp_kses_post( nl2br( $subscription->get_payment_method_to_display() ) ); // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
```
You'll notice as part of this PR that there are a couple of additional changes to capitalisation and removing `:`. This is so these labels are consistant with other labels in this section. 

eg 

| Before (title case) | After (sentence case) |
|--------|--------|
| <img width="331" alt="Screenshot 2023-10-31 at 1 25 22 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/69346809-988e-4ffa-80bd-962d577a9404"> | <img width="326" alt="Screenshot 2023-10-31 at 1 26 46 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/857cdcc2-9945-4dc4-a254-1b89b0451c45"> |



| Before | After |
|--------|--------|
| <img width="667" alt="Screenshot 2023-10-31 at 1 31 29 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/1bb8c88a-fdeb-4b7c-84ad-c0ee6174b4e7"> <p align="center"><sup>Title case and erroneous `:`</p><sup> | <img width="684" alt="Screenshot 2023-10-31 at 1 29 19 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/b508c48b-ef6d-4873-aa63-c19563e27a18"> |

james-allan committed:
Fix phpcs

james-allan commented on includes/class-wcs-change-payment-method-admin.php:
```diff
@@ -91,12 +90,24 @@ public static function display_fields( $subscription ) {
+				'wc-subscription-auto-renew',
+				[
+					'type'    => 'checkbox',
+					'label'   => esc_html__( 'Auto renew', 'woocommerce-subscriptions' ) . wcs_help_tip( __( 'This subscription has automatic renewals turned off. Use this option to turn them on.', 'woocommerce-subscriptions' ) ),
```
This line includes the help tip which looks like this

<img width="400" alt="Screenshot 2023-10-31 at 1 38 07 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/048262d0-4461-4ded-a1be-72ef8dc6d80d">

Mayisha reviewed (commented):
The fix works as described. 
I have one question regarding the newly added checkbox. As the reported issue describes, after changing the toggle from `My account` page, the payment method can not be changed from `Edit subscription` page by admin. Instead of fixing the payment method dropdown so that the selected gateway is saved, why are we introducing this new checkbox? :thinking:

james-allan commented:
> after changing the toggle from My account page, the payment method can not be changed from Edit subscription page by admin

The payment method can be changed from the edit subscription page after using the toggle from the my account page. My understanding of that part was that it can be changed but not saved. I've included a video of what I believe they were referring to below. 

https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/448a43a5-19cc-4eb3-9600-fd0e75fa944d

The reason changing the payment method in this video isn't "sticking" is because the subscription has the "requires manual renewal" flag set which means irrespective of the payment method, it is locked to manual renewals. 

Adding this checkbox allows the user to turn off that manual renewal lock and therefore resume automatic renewals and for the chosen payment method to take effect. 

Another approach we could look into is something like this: 

1. The payment method dropdown is set to manual renewal if the requires manual renewal flag is on. 
2. If admin change the payment method and save, we assume they intend on turning on automatic renewals and so switch it to automatic renewals. 

I'm trying to think if that would break any flows, making an assumption like that. What do you think though @mayisha, would that make more sense rather than a checkbox?

james-allan committed:
Select the "Manual Renewal" option by default if the subscription is manual

Prior to this change the chosen option would be whatever the subscriptions payment method
in the database was even if it was manual. This was confusing but also prevented
admin from turning off manual renewals.

james-allan committed:
Remove the auto renew checkbox

james-allan committed:
Turn off manual renewals if the subscriptions payment method is changed to an automatic payment method

james-allan commented:
> What do you think though @Mayisha, would that make more sense rather than a checkbox?

I've been having a go at implementing this and I think this makes a lot more sense than the checkbox. 

https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/1f332784-f039-4d8e-98fe-2da69138776b

I think this is inline with what you were suggesting right?

james-allan committed:
Update changelog entry

Mayisha commented:
Sorry for not elaborating on my question before :no_mouth: 

This implementation https://github.com/Automattic/woocommerce-subscriptions-core/pull/535#issuecomment-1791853080 exactly aligns with what I meant. :100:
